### PR TITLE
printer block right clicking issue

### PR DESCRIPTION
**Pull request by coehlrich**
i think i have fixed the printer block right clicking (it should fix issue #110 ).

### CODE REVIEW

File: src/main/java/com/github/lunatrius/schematica/client/printer/SchematicPrinter.java [hunk 1]
```diff
@@ -194,7 +194,7 @@ private boolean placeBlock(final WorldClient world, final EntityPlayerSP player,
             return false;
         }
 
-        if (ConfigurationHandler.destroyBlocks && !world.isAirBlock(realPos) && this.minecraft.playerController.isInCreativeMode()) {
+        if (ConfigurationHandler.destroyBlocks && !world.isAirBlock(realPos) && this.minecraft.playerController.isInCreativeMode() && PlacementRegistry.INSTANCE.getPlacementData(blockState, itemStack).getExtraClicks(blockState) == 0) {
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/printer/SchematicPrinter.java]: What exactly is this change supposed to do? Doesn't really have anything to do with the original issue...

File: src/main/java/com/github/lunatrius/schematica/client/printer/SchematicPrinter.java [hunk 2]
```diff
@@ -332,6 +332,7 @@ private boolean placeBlock(final WorldClient world, final EntityPlayerSP player,
         success = placeBlock(world, player, itemStack, offset, side, hitVec, hand);
         for (int i = 0; success && i < extraClicks; i++) {
             success = placeBlock(world, player, itemStack, offset, side, hitVec, hand);
+            Thread.sleep(50);
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/printer/SchematicPrinter.java]: No. This is not even a temporary solution to the real issue and only works for the default settings. Try setting the placement delay to minimum with instant placement enabled.

Whenever you'll move around an area with lots of blocks that require multiple clicks the entire game will freeze for `50 * extra clicks` milliseconds.

File: src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java [hunk 3]
```diff
@@ -47,7 +47,7 @@
             final boolean isPlaced = BlockStateHelper.areBlockStatesEqual(blockState, mcBlockState);
 
             ItemStack stack = null;
-
+
```
Comment (Lunatrius) [inline src/main/java/com/github/lunatrius/schematica/client/util/BlockList.java]: This shouldn't be in the PR at all.

**Comment by coehlrich** [comment]
is there anything that isnt said in the original issue?

**Comment by coehlrich** [comment]
is there anything that isnt said in the original issue?

**Comment by Lunatrius** [comment]
Don't think so. But a large chunk of the printer code has to be rewritten to use a queue system instead of instant commands, taking into account the player's movement , extra clicks and possibly other things that i can't remember right now.

**Comment by coehlrich** [comment]
where do the amount of extra clicks get calculated?

**Comment by Lunatrius** [comment]
Inside the extra clicks method?

**Comment by coehlrich** [comment]
the last time i looked at the Iextraclicks file i could only find one method and that was to call another method somewhere.

**Comment by Lunatrius** [comment]
Why does it even matter? It simply takes the block state and checks how many clicks have to be performed.

**Comment by coehlrich** [comment]
i am going to need to change it a bit i am going to need it to also it subtract the amount of clicks done from the total amount of clicks

**Comment by Lunatrius** [comment]
It already does that. It takes the block state as an argument, checks how many clicks have been already made and returns the amount of extra clicks that have to be done still.

But as I have already said it, this is **not** a simple thing to fix. You can't just slap a random function call here and there and make it magically work for a random subset of possible scenarios (like the sleep hack). It **must** be implemented as a proper priority queue to apply all the required actions (and make sure that they are still valid at that point, because the player can move around...).

**Comment by coehlrich** [comment]
what changes do i need to make to the printer overall? (also i am doing them on a different branch than the one hat is currently linked to this pull request)

**Comment by coehlrich** [comment]
i will have to open a new pull request to change the branch

**Comment by Lunatrius** [comment]
> But a **large chunk** of the printer code has to be **rewritten** to use a queue system instead of instant commands, taking into account the player's movement, extra clicks and possibly other things that I can't remember right now.

**Comment by coehlrich** [comment]
also it is the master branch not the repeater right clicking branch i would change the branch for this pull request but i found out the only way to do that is to start a new pull request.

**Comment by Lunatrius** [comment]
I know, I have already [looked at it](https://github.com/Lunatrius/Schematica/compare/master...coehlrich:master). Those changes are meaningless because most of those methods would have to be remove/moved to another class to encapsulate the data...

**Comment by coehlrich** [comment]
i have put it in another class.

**Comment by Lunatrius** [comment]
*Please, save your time and effort (and mine, while trying to respond to this PR) and let me do this once I have the time.* I appreciate the willingness to help out and to contribute to the development of the mod, but you said it yourself in the other pull request: you have close to no Java knowledge (I assume you're in the process of learning it).

You didn't really understand my statement.

> ... because most of those method**s** would have to be...

The **entire** printer class has to be **rewritten/refactored** with the already mentioned changes that have to be made to make it work properly with a queue system that has yet to be implemented.

The changes that you have done so far do nothing at all. The placement range check is already performed earlier in the call chain and adds nothing but overhead with that implementation.

**Comment by NatoBoram** [comment]
Wouldn't it be better to close this PR from  2016?

**Pull request closed without merging**